### PR TITLE
Key change menu: Fix empty string translation

### PR DESCRIPTION
**Pull request by minduser00**
Fix for incorrect translation of empty strings in key change menu

In the key change menu, when a button key not have name an empty string is passed to gettext.
The empty string is reserved for gettext to return de header of the .po file an this is shoved in the button

minetest-es

The capture is from version 0.4.16 but the code is the same in the master branch.

### CODE REVIEW

File: src/gui/guiKeyChangeMenu.cpp [hunk 1]
```diff
@@ -153,9 +153,8 @@ void GUIKeyChangeMenu::regenerateGui(v2u32 screensize)
 		{
 			core::rect < s32 > rect(0, 0, 100, 30);
 			rect += topleft + v2s32(offset.X + 120, offset.Y - 5);
-			const wchar_t *text = wgettext(k->key.name());
-			k->button = Environment->addButton(rect, this, k->id, text);
-			delete[] text;
+			k->button = Environment->addButton(rect, this, k->id,
+							   k->key.name()[0] ? wgettext(k->key.name()) : L"");
```
Comment (SmallJoker) [inline src/gui/guiKeyChangeMenu.cpp]: How about deleting the pointer from wgettext? How about this:
```C++
const wchar_t *text = wgettext(k->key.name());
k->button = Environment->addButton(rect, this, k->id,
	k->key.name()[0] ? text : L"");
delete[] text;
```

Comment (sfan5) [reply src/gui/guiKeyChangeMenu.cpp]: `L""` is a static string in that case, you can't delete it

Comment (minduser00) [reply src/gui/guiKeyChangeMenu.cpp]: I tried it yesterday with L"" and crash wen key change menu opens:

guadalinex@guadalinex:~/minetest$ bin/minetest
2018-03-27 20:36:09: ERROR[Main]: Subgame specified in default_game [minetest] is invalid.
Loaded texture: /home/guadalinex/minetest/games/minimal/menu/icon.png
Loaded texture: /home/guadalinex/minetest/textures/base/pack/menu_header.png
*** Error in `bin/minetest': munmap_chunk(): invalid pointer: 0x0852f2b0 ***
Abortado

Whit *L""  works in my linux but not in other systems, can't compile, so the only option is with an "if/else" or suppresing the "text" variable

**Comment by minduser00** [comment]
This is the other option with the variable:

```
			if (k->key.name()[0]) {
				const wchar_t *text = wgettext(k->key.name());
				k->button = Environment->addButton(rect, this, k->id, text);
				delete[] text;
			} else {
				k->button = Environment->addButton(rect, this, k->id, L"");
			}

```

**Comment by paramat** [comment]
Same issue as #7088

File: src/gui/guiKeyChangeMenu.cpp (hunk 1 above)
Comment (minduser00) [reply src/gui/guiKeyChangeMenu.cpp]: Thanks @SmallJoker you were right. I had to delete the pointer returned by wgettext as indicated in src/gettext.h

**Comment by minduser00** [comment]
Yes, it is possible to modify the wgettext and strgettext functions to avoid that problem with the translation of a null string, but that will avoid the legitimate possibility, which is now present, of using wgettext(""") to obtain the information of the translation file. Are you sure you want to do that?
If you want it that way, it would also solve #7088

File: src/gettext.h [hunk 2]
```diff
@@ -51,10 +51,19 @@ extern wchar_t *utf8_to_wide_c(const char *str);
 // The returned string is allocated using new
 inline const wchar_t *wgettext(const char *str)
 {
-	return utf8_to_wide_c(gettext(str));
+#if USE_GETTEXT
```
Comment (SmallJoker) [inline src/gettext.h]: Unnecessary use of preprocessors here. `gettext()` will functionally be ignored if gettext is unavailable. (See L38)

Comment (minduser00) [reply src/gettext.h]: @SmallJoker 
Yes, the gettext function will be ignored if it is compiled without gettext, I known the use of:
```#define gettext(String) String```
But why check if the string is empty if it is not going to be translated? 
Without gettext when compiling would be left:   
```
    return str[0] ? utf8_to_wide_c(str) : utf8_to_wide_c("");
```
and
```
    return text.empty() ? text.c_str() : text.c_str();
```
In both cases they'd only waste time testing a character for nothing, it's a bit ugly, isn't it?

Comment (SmallJoker) [reply src/gettext.h]: It's about the readability. The time spent for a conditional jump is negligible in this case.

Comment (minduser00) [reply src/gettext.h]: Ok

**Comment by sfan5** [comment]
9a06d6aa9fa2676a3fb4fa494f9318d0c5d7becb

**Comment by minduser00** [comment]
I warn you of these crash sfan5 & you whites 13 days ago than can't be use L""
:yum:

**Pull request closed without merging**